libs/community/langchain_community/tools/databricks/tool.py

```diff
@@ -0,0 +1,230 @@
+            return Dict[str, value_type]  # type: ignore
+        elif type == "struct":
+            fields = {}
+            for field in uc_type_json["fields"]:
+                field_type = uc_type_to_pydantic_type(field["type"])
+                if field.get("nullable"):
+                    field_type = Optional[field_type]  # type: ignore
+                comment = (
+                    uc_type_json["metadata"].get("comment")
+                    if "metadata" in uc_type_json
+                    else None
+                )
+                fields[field["name"]] = (field_type, Field(..., description=comment))
+            uc_type_json_str = json.dumps(uc_type_json, sort_keys=True)
+            type_hash = md5(uc_type_json_str.encode()).hexdigest()[:8]
+            return create_model(f"Struct_{type_hash}", **fields)  # type: ignore
+        else:
+            raise TypeError(f"Unknown type {uc_type_json}. Try upgrading this package.")
+
+
+def generate_args_schema(function: FunctionInfo) -> Optional[Type[BaseModel]]:
+    if function.input_params is None:
+        return None
+    params = function.input_params.parameters
+    assert params is not None
+    fields = {}
+    for p in params:
+        assert p.type_json is not None
+        type_json = json.loads(p.type_json)["type"]
+        pydantic_type = uc_type_to_pydantic_type(type_json)
+        description = p.comment
+        default: Any = ...
+        if p.parameter_default:
+            pydantic_type = Optional[pydantic_type]  # type: ignore
+            default = None
+            # TODO: Convert default value string to the correct type.
+            # We might need to use statement execution API
+            # to get the JSON representation of the value.
+            default_description = f"(Default: {p.parameter_default})"
+            if description:
+                description += f" {default_description}"
+            else:
+                description = default_description
+        fields[p.name] = (
+            pydantic_type,
+            Field(default=default, description=description),
+        )
+    return create_model(
+        f"{function.catalog_name}__{function.schema_name}__{function.name}__params",
+        **fields,  # type: ignore
+    )
+
+
+def get_tool_name(function: FunctionInfo) -> str:
+    tool_name = f"{function.catalog_name}__{function.schema_name}__{function.name}"[
+        -64:
+    ]
+    return tool_name
+
+
+class UCFunctionTool(BaseTool):
+    """Tool for calling a user-defined function in Unity Catalog."""
+
+    function: FunctionInfo
+    warehouse_id: str
+    workspace_client: WorkspaceClient
+
+    def __init__(
+        self,
+        function_name: str,
+        warehouse_id: str,
+        workspace_client: Optional[WorkspaceClient] = None,
+        **kwargs: Any,
+    ):
+        if workspace_client is None:
+            workspace_client = WorkspaceClient()
+        function = workspace_client.functions.get(function_name)
```

Comment:
Can we move this loading logic into the `UCFunctionTools.get_tool` factory function and make the init method just a setter?

Reply:
I would also be supportive of not implementing `UCFunctionTool` as a class and instead having it be a function that returns `Tool(...)` e.g. https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/agent_toolkits/load_tools.py#L180

I think having a single entry point of UCFunctionTools is more intuitive

Reply:
As a quick fix could also just prefix this class with an underscore, so we're more flexible to change it later